prometheus/promhttp/instrument_server.go

```diff
@@ -0,0 +1,363 @@
+			label = sanitizeCode(delegate.status)
+		}
+		obs.WithLabelValues(label).Observe(time.Since(now).Seconds())
+	})
+}
+
+// Latency accepts an ObserverVec interface and an http.Handler, returning a
+// new http.Handler that wraps the supplied http.Handler. The provided
+// CounterVec must be registered in a registry in order to be used.
+// If the wrapped http.Handler has not set a status code, i.e. the value is
+// currently 0, the supplied counter will report a 200.
+// The instance labels "code" and "method" are supported on the provided
+// CounterVec.
+func Counter(counter *prometheus.CounterVec, next http.Handler) http.Handler {
+	var (
+		code   bool
+		method bool
+		err    error
+	)
+
+	// TODO(beorn7): Remove this hacky way to check for instance labels
+	// once Descriptors can have their dimensionality queried.
+	if _, err = counter.GetMetricWith(prometheus.Labels{"code": "0", "method": "get"}); err == nil {
+		code = true
+		method = true
+	} else if _, err = counter.GetMetricWith(prometheus.Labels{"method": "get"}); err == nil {
+		method = true
+	} else if _, err = counter.GetMetricWith(prometheus.Labels{"code": "0"}); err == nil {
+		code = true
+	} else if _, err = counter.GetMetricWith(prometheus.Labels{}); err != nil {
+		panic(`counter middleware requires instance labels "code", "method", or both "code" and "label".`)
+	}
+	// TODO: There is no delete exposed on interface Counter.
+	// c.Delete(labels)
+
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		delegate := &responseWriterDelegator{ResponseWriter: w}
+		labels := prometheus.Labels{}
+
+		next.ServeHTTP(delegate, r)
+
+		if code {
+			labels["code"] = sanitizeCode(delegate.status)
+		}
+		if method {
+			labels["method"] = sanitizeMethod(r.Method)
+		}
+		counter.With(labels).Inc()
+	})
+}
+
+// RequestSize accepts an Observer interface and an http.Handler, returning a
+// new http.Handler that wraps the supplied http.Handler. The provided Observer
+// must be registered in a registry in order to be used.
+func RequestSize(obs prometheus.Observer, next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		next.ServeHTTP(w, r)
+		size := computeApproximateRequestSize(r)
+		obs.Observe(float64(size))
+	})
+}
+
+// ResponseSize accepts an Observer interface and an http.Handler, returning a
+// new http.Handler that wraps the supplied http.Handler. The provided Observer
+// must be registered in a registry in order to be used.
+func ResponseSize(obs prometheus.Observer, next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		delegate := &responseWriterDelegator{ResponseWriter: w}
+
+		_, cn := w.(http.CloseNotifier)
+		_, fl := w.(http.Flusher)
+		_, hj := w.(http.Hijacker)
+		_, ps := w.(http.Pusher)
+		_, rf := w.(io.ReaderFrom)
+		var rw http.ResponseWriter
+		if cn && fl && hj && rf && ps {
+			rw = &fancyResponseWriterDelegator{delegate}
```

Comment:
We actually should check for the interface upgrade in the other delegator uses above, too.

This should be helper code in a function used by everybody else.